### PR TITLE
Fix sample code on InputControl docs

### PR DESCRIPTION
**Pull request by flexseth**
## What?
InputControl code samples broken - the `useState` package has moved from `wordpress/compose` to `wordpress/element`

## How?
When trying to use an InputControl, I found that many of the Gutenberg package imports had been updated, but not those for the InputControl component. This affects the [Developer Documentation](https://developer.wordpress.org/block-editor/reference-guides/components/input-control/#usage) and also [Storybook](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-inputcontrol--docs). 

There were three declarations that needed to be changed in the Gutenberg folder. Search & Replace in VS Code worked to fix the code and this PR represents the updated Gutenberg codebase with changes to the developer documentation. 

## Testing Instructions

1. `npx @wordpress/create-block input-control-block`
2. Start the build to watch for changes
3. Activate the Input Control Block plugin
4. Add an input control block to a new post
5. Modify the edit function with the [sample code](https://developer.wordpress.org/block-editor/reference-guides/components/input-control/#usage), see steps below
6. Add the import headers above the Edit() function
```js
import { __experimentalInputControl as InputControl } from '@wordpress/components';
import { useState } from '@wordpress/compose';
```
7. Change the Edit function to match sample code
```js 
export default function Edit() {
	const [ value, setValue ] = useState( '' );
	
	return (
		<InputControl
			value={ value }
			onChange={ ( nextValue ) => setValue( nextValue ?? '' ) }
		/>
	);
};
```

8. At this point the block breaks
Screenshot 2024-03-01 at 5 10 47 PM
console error shown below

Screenshot 2024-03-01 at 5 11 16 PM

### Testing the broken interface

1. In Edit.js
**change**
```js
import { useState } from '@wordpress/compose';
```
**to**
```js
import { useState } from '@wordpress/element';
```
3. Save changes
4. Refresh the page or post where you inserted the block
6. The block should now show an input box
Screenshot 2024-03-01 at 5 25 18 PM

## About this PR
This PR simply changes the location of the package imports, so no one else runs into this when trying to work with the InputControl component.

### Screenshots included
- UI when not working
- Error code
- Functional UI

### CODE REVIEW

**Comment by t-hamano** [comment]
Thanks for the PR! Looking at #54908, it seems that it is currently recommended to use `react` instead of `@wordpress/element` 🤔

cc @youknowriad @tyxla

**Comment by flexseth** [comment]
> Thanks for the PR! Looking at #54908, it seems that it is currently recommended to use `react` instead of `@wordpress/element` 🤔
> 
> cc @youknowriad @tyxla

Thank you for bringing this up @t-hamano! I have seen the packages imported this way and never thought about the distinction until just now. 

#### More info on the changes in React abstraction
- https://github.com/WordPress/gutenberg/pull/54494

#### Possible reason to use `wordpress/element`
There may be some [additional controls](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-inputcontrol--docs) offered for the InputControl component over the native React.js `input` component ([see attributes](https://react.dev/reference/react-dom/components/input)), such as
- `prefix`
- `suffix`
- `_next40pxDefaultSize`

#### Explanation of controls
- set text to display before input field
- set text to display after input field
- make the input a little larger 

#### Testing these controls with `react` import value
I will do some testing to see if I can use these controls without importing the package from Gutenberg.

**Review by flexseth — COMMENTED** [review]
Checked InputControl component with 
`import {useState} from 'react'`

This abstraction works. There may be additional features of useState that are advanced and customized for the `useState` hook for WordPress. If this is the case, the block will break and the user will have to import the package directly from WordPress to use this functionality. This is a known limitation per
- https://github.com/WordPress/gutenberg/issues/54074
> Note that we do have some functions in @wordpress/element that are additional like renderToString or createInterpolateElement...

**Comment by youknowriad** [comment]
There's strictly no difference between `@wordpress/element` and `react` for these React APIs. I think we should just make sure that we're consistent in our code base.

It seems that right now:

 - For documentation purpose, we use `react` 
 - For code, we're still using `@wordpress/element`

So let's just maintain that status-quo for now. which I guess mean ship this PR.

File: packages/components/src/input-control/index.tsx [hunk 1]
```diff
@@ -126,7 +126,7 @@ export function UnforwardedInputControl(
  *
  * ```jsx
  * import { __experimentalInputControl as InputControl } from '@wordpress/components';
- * import { useState } from '@wordpress/compose';
+ * import { useState } from '@wordpress/element';
```
Comment (youknowriad) [inline packages/components/src/input-control/index.tsx]: This should be "react" I think.

File: packages/components/src/input-control/index.tsx [hunk 2]
```diff
@@ -8,7 +8,7 @@ import type { ForwardedRef } from 'react';
  * WordPress dependencies
  */
 import { useInstanceId } from '@wordpress/compose';
-import { useState, forwardRef } from '@wordpress/element';
+import { useState, forwardRef } from 'react';
```
Comment (youknowriad) [inline packages/components/src/input-control/index.tsx]: This should be @wordpress/element

File: packages/components/CHANGELOG.md [hunk 3]
```diff
@@ -3,7 +3,7 @@
 ## Unreleased
 
 ### Bug Fix
-
+-   `InputControl`: Fix sample code on InputControl docs [#2610](https://github.com/WordPress/gutenberg/pull/59517)
```
Comment (tyxla) [inline packages/components/CHANGELOG.md]:
```suggestion
-   `InputControl`: Fix sample code on InputControl docs [#59517](https://github.com/WordPress/gutenberg/pull/59517)
```

**Review by tyxla — COMMENTED** [review]
> It seems that right now:
> 
> For documentation purpose, we use react
> For code, we're still using @wordpress/element

Riad put it really well. On top of that, other than another minor CHANGELOG correction, think this should be good to go - thanks!

**Comment by flexseth** [comment]
Thank you @youknowriad, @t-hamano and @tyxla for support! I will update and squash the commits to clean this up. This has been a great exercise in understanding commit history and the upstream flow, so I really do appreciate your feedback and direction!

File: packages/components/src/input-control/index.tsx (hunk 1 above)
Comment (flexseth) [reply packages/components/src/input-control/index.tsx]: Changed import declaration to `react`. 

This is sample code in the documentation, the code is wrapped in Markdown syntax representing JSX
```
```jsx
```

[see notes on codebase consistency](https://github.com/WordPress/gutenberg/pull/59517#:~:text=our%20code%20base.-,It%20seems%20that%20right%20now%3A,-For%20documentation%20purpose)

File: packages/components/src/input-control/index.tsx (hunk 2 above)
Comment (flexseth) [reply packages/components/src/input-control/index.tsx]: Reverted to `@wordpress/element` import. 

The context here is source code ([see notes on codebase consistency](https://github.com/WordPress/gutenberg/pull/59517#:~:text=we%20use%20react-,For%20code%2C%20we%27re%20still%20using%20%40wordpress/element,-So%20let%27s%20just))

File: packages/components/CHANGELOG.md (hunk 3 above)
Comment (flexseth) [reply packages/components/CHANGELOG.md]: CHANGELOG has been updated: Link to PR fixing issue #59517

**Comment by flexseth** [comment]
Thank you for the support on this @t-hamano.
 
> Successfully rebased and updated refs/heads/docs/update-code-samples.

- Commits rebased
- Followed message standards for final commit
- Some checks are not passing but are related to other packages - `useEffect` - mostly

Thanks again! This was by far the most challenging software thing I've ever wrestled with 😄

Requesting final review @tyxla - thanks for the help on the CHANGELOG update.

File: packages/components/src/input-control/index.tsx (hunk 1 above)
Comment (t-hamano) [reply packages/components/src/input-control/index.tsx]: It has already been mentioned, but isn't it "react" instead of "element"?

**Comment by flexseth** [comment]
@tyxla - I tried to change the commit message and somehow added quite a few more commits. Please hold off on reviewing. I'll get some help from my mentor after they return from WC Asia.

File: packages/components/src/input-control/index.tsx (hunk 1 above)
Comment (flexseth) [reply packages/components/src/input-control/index.tsx]: I will look into it. It's possible when I was making changes it wasn't syncing to the correct branch. 

Still getting familiar with the workflow of Git and contributing upstream vs. local working copy.

Will report back, @t-hamano

**Comment by youknowriad** [comment]
Don't worry too much about commit messages, the person merging the PR will change and adapt it properly.

File: packages/components/src/input-control/index.tsx (hunk 2 above)
Comment (youknowriad) [reply packages/components/src/input-control/index.tsx]: This is still using "react" though? Unless GitHub is not showing me the latest changes :)

File: packages/components/src/input-control/index.tsx (hunk 1 above)
Comment (flexseth) [reply packages/components/src/input-control/index.tsx]: This should be resolved now.

File: packages/components/CHANGELOG.md [hunk 4]
```diff
@@ -3,7 +3,7 @@
 ## Unreleased
 
 ### Bug Fix
-
+-    `InputControl`: Fix sample code on InputControl docs [#59517](https://github.com/WordPress/gutenberg/pull/59517)
```
Comment (t-hamano) [inline packages/components/CHANGELOG.md]:
```suggestion
-   `InputControl`: Fix sample code on InputControl docs [#59517](https://github.com/WordPress/gutenberg/pull/59517)
```

**Review by t-hamano — APPROVED** [review]
LGTM!  I only deleted unnecessary spaces in CHANGELOG.

**Comment by flexseth** [comment]
Thank you everyone for helping out on this!

**Pull request merged**